README.md
# Simple Database
### Used personally for many different NodeJS projects.
## To Do List
- [ ] Publish to NPM
- [ ] Update/optimise server database
- [ ] Streamline main framework further

changelog.md
1.0.1 / 2022-01-17
===================

    * Fix: issues with values not having the properties they should
    * Update: the database function get() to now return a object with a custom prototype with functions related to the query

index.js
const http = require("http"),
    https = require("https");
const Database = function(token, url) {
    if (!token) throw new TypeError("database.construct() valid token must be provided");
    const parsedURL = url ? new URL(url) : { "protocol": "https:", "port": 443 };
    this.connection = { hostname: parsedURL["hostname"] ? parsedURL["hostname"] : "database.macstudio.pro" }
    this.connection.protocol = parsedURL["protocol"] === "https:" ? https : http;
    this.connection.token = token;
    if (parsedURL["port"]) this.connection.pport = parsedURL["port"];
}
/**
 * make a request to any server
 * @param {Object} options - options for the request
 * @param {Any} data - the data to send with the request
 */
Database.prototype._request = function(options, data) {
    const type = data ? typeof data.on === 'function' : false;
    return new Promise(async (resolve, reject) => {
        try {
            const headers = options["headers"] ? options["headers"] : {};
            headers["authorization"] = this.connection.token;
            options["headers"] = headers;
            options["hostname"] = this.connection.hostname;
            if (this.connection.port) options["port"] = this.connection.port;
            if (type) options["headers"]["Transfer-Encoding"] = "chunked";
            const req = this.connection.protocol.request(options, async (res) => {
                resolve(res);
            });
            req.on('error', error => reject(error));
            if (data) {
                if (type) {
                    data.pipe(req, { end: true });
                } else {
                    req.write(data);
                    req.end();
                }
            } else {
                req.end();
            }
        } catch (e) {
            reject(e);
        }
    });
}
/**
 * get a value from the database
 * @param {String} key - the key of the data to return
 * @returns {Object} - the value stored within the database
 */
Database.prototype._get = function(key) {
    return new Promise(async (resolve, reject) => {
        try {
            const res = await this._request({
                path: `/content/${encodeURIComponent(key)}`,
                method: 'GET',
            });
            if (res.statusCode === 200) {
                var data = [];
                res.on("data", async (chunk) => {
                    data.push(chunk);
                }).on("end", async () => {
                    data = Buffer.concat(data);
                    resolve(JSON.parse(data));
                });
            } else if (res.statusCode === 404) {
                resolve({});
            } else {
                throw { status: res.statusCode };
            }
        } catch (e) {
            reject(e);
        }
    });
}
/**
 * sets a value into the database
 * @param {String} key - the key to store the value
 * @param {String} value - the value to save
 */
Database.prototype._set = function(key, value) {
    return new Promise(async (resolve, reject) => {
        try {
            const res = await this._request({
                path: `/content/${encodeURIComponent(key)}`,
                method: 'PUT',
            }, value);
            if (res.statusCode === 200) {
                resolve();
            } else {
                throw { status: res.statusCode };
            }
        } catch (e) {
            reject(e);
        }
    });
}
/**
 * deletes a value from the database
 * @param {String} key - the key for the value to delete
 */
Database.prototype._delete = function(key) {
    return new Promise(async (resolve, reject) => {
        try {
            const res = await this._request({
                path: `/content/${encodeURIComponent(key)}`,
                method: 'DELETE',
            });
            if (res.statusCode === 200) {
                resolve();
            } else {
                throw { status: res.statusCode };
            }
        } catch (e) {
            reject(e);
        }
    });
}
/**
 * lists all the files in a path
 * @param {string} key Key
 * @returns {object} Directory list
 */
Database.prototype.list = async function(key) {
    return new Promise(async (resolve, reject) => {
        try {
            const res = await this._request({
                path: `/list/${encodeURIComponent(key)}`,
                method: 'GET',
            });
            if (res.statusCode === 200) {
                const data = [];
                res.on("data", async (chunk) => {
                    data.push(chunk);
                }).on("end", async () => {
                    resolve(JSON.parse(Buffer.concat(data)));
                });
            } else {
                reject(res.statusCode);
            }
        } catch (e) {
            reject(e);
        }
    });
}
/**
 * returns object to manipulate another object
 * @param {string} key key
 * @param {object} defaults default values
 * @returns {object} data
 */
Database.prototype.get = async function(key, defaults) {
    return new Promise(async (resolve, reject) => {
        const prototype = {
            "_key": key,
            "_database": this,
            "_save": async function() {
                return new Promise(async (resolve, reject) => {
                    try {
                        resolve(await this._database._set(this._key, JSON.stringify(this)));
                    } catch (e) {
                        reject(e);
                    }
                });
            },
            "_delete": async function() {
                return new Promise(async (resolve, reject) => {
                    try {
                        resolve(await this._database._delete(this._key));
                    } catch (e) {
                        reject(e);
                    }
                });
            },
        };
        const data = await this._get(key);
        const item = Object.create(prototype);
        Object.assign(item, defaults ? defaults : {});
        Object.assign(item, data);
        resolve(item);
    });
}

exports = module.exports = Database;

archives-db/databasex.js
const fs = require("fs");
const http = require("http");
const https = require("https");


class Network {
    async uploadData(url, data) {
        const type = typeof this.value.on === 'function';
        return new Promise(async (resolve, reject) => {
            try {
                const options = {
                    hostname: this.connection.hostname,
                    //port: this.connection.port,
                    path: `/content/${encodeURIComponent(url)}`,
                    method: 'PUT',
                    headers: {
                        "authorization": this.connection.authorization,
                    }
                }
                if (type) options["headers"]["Transfer-Encoding"] = "chunked";
                const req = this.connection.protocol.request(options, res => {
                    if (res.statusCode != 500) {
                        resolve();
                    } else {
                        reject(500);
                    }
                });
                req.on('error', error => {
                    reject(error);
                });
                if (type) {
                    data.pipe(req, { end: true });
                } else {
                    req.write(data);
                    req.end();
                }
            } catch (e) {
                reject(e);
            }
        });
    }
    async getData(url) {
        return new Promise(async (resolve, reject) => {
            try {
                const req = this.connection.protocol.request({
                    hostname: this.connection.hostname,
                    //port: this.connection.port,
                    path: `/content/${encodeURIComponent(url)}`,
                    method: 'GET',
                    headers: {
                        "authorization": this.connection.authorization,
                    }
                }, async res => {
                    return resolve(res); // get statuscode res.statusCode
                });
                req.on('error', error => {
                    reject(error);
                })
                req.end()
            } catch (error) {
                reject(error);
            }
        });
    }
    async listData(url) {
        return new Promise(async (resolve, reject) => {
            try {
                const req = this.connection.protocol.request({
                    hostname: this.connection.hostname,
                    //port: this.connection.port,
                    path: `/list/${encodeURIComponent(url)}`,
                    method: 'GET',
                    headers: {
                        "authorization": this.connection.authorization,
                    }
                }, async res => {
                    return resolve(res); // get statuscode res.statusCode
                });
                req.on('error', error => {
                    reject(error);
                })
                req.end()
            } catch (error) {
                reject(error);
            }
        });
    }
    async deleteData(url) {
        return new Promise(async (resolve, reject) => {
            try {
                const req = this.connection.protocol.request({
                    hostname: this.connection.hostname,
                    //port: this.connection.port,
                    path: `/content/${encodeURIComponent(url)}`,
                    method: 'DELETE',
                    headers: {
                        "authorization": this.connection.authorization,
                    }
                }, async res => {
                    return resolve(res); // get statuscode res.statusCode
                });
                req.on('error', error => {
                    reject(error);
                })
                req.end()
            } catch (error) {
                reject(error);
            }
        });
    }
}

class DatabaseObject extends Network {
    // what will be returned when calls the get("name") function on a database
    constructor(url, connection) {
        super()
        this.url = url;
        this.value = undefined;
        this.connection = connection;
    }
    async fetch() {
        return new Promise(async (resolve, reject) => {
            try {
                this.value = await this.fetchValue();
                resolve();
            } catch (e) {
                reject(e);
            }
        });
    }
    async delete() {
        this.value = undefined;
        this.deleteData(this.url);
    }
    async save() {
        return new Promise(async (resolve, reject) => {
            try {
                await this.uploadData(this.url, this.url.endsWith(".json") ? JSON.stringify(this.value) : this.value);
                resolve();
            } catch (e) {
                console.log(e);
                reject(e);
            }
        });
    }
    async fetchValue() {
        return new Promise(async (resolve, reject) => {
            try {
                const data = await this.getData(this.url);
                if (data.statusCode === 404) {
                    return resolve(undefined);
                } else if (data.statusCode != 200) {
                    throw Error(data.statusCode);
                }
                if (this.url.endsWith(".json")) {
                    var contents = [];
                    data.on("data", function (d) {
                        contents.push(d);
                    }).on("end", function () {
                        resolve(JSON.parse(Buffer.concat(contents).toString()));
                    });
                } else if (this.url.endsWith(".txt")) {
                    var toString = [];
                    data.on("data", function (d) {
                        toString.push(d);
                    }).on("end", function () {
                        resolve(Buffer.concat(toString).toString());
                    });
                } else {
                    resolve(data);
                }
            } catch (e) {
                reject(e);
            }
        });
    }
}

class Database extends Network {
    constructor() {
        super();
        this.connection = {};
    }
    async connect(connectionString, authorization) {
        return new Promise(async (resolve, reject) => {
            // initiate a connection with the server and set this variables
            try {
                const parseConnection = new URL(connectionString);
                if (!parseConnection.hostname || !parseConnection.protocol) {
                    throw Error("Does not contain needed params in connection URL.");
                }
                this.connection.hostname = parseConnection.hostname;
                this.connection.protocol = parseConnection.protocol === "https:" ? https : http;
                // this.connection.port = parseConnection.port || parseConnection.protocol === "https:" ? 445 : 80;
                if (!authorization) {
                    throw Error("Missing authorization token.");
                }
                this.connection.authorization = authorization;
                resolve();
            } catch (e) {
                reject(e);
            }
        });
    }
    get(url) {
        // returns an object from the database
        return new DatabaseObject(url, this.connection);
    }
    async list(url) {
        return new Promise(async (resolve, reject) => {
            try {
                const list = await this.listData(url);
                const data = [];
                list.on("data", function(d) {
                    data.push(d);
                }).on("end", function() {
                    resolve(JSON.parse(Buffer.concat(data).toString()));
                });
            } catch(e) {
                reject(e);
            }
        });
    }
}

module.exports = Database;